File: Algorithms/src/3666. Minimum Operations to Equalize Binary String/Code/solution.cpp
class Solution
{
public:
    int minOperations(string s, int k)
    {
        int n = s.length();
        int zero = 0;

        for (char c : s)
            if (c == '0')
                zero++;

        if (zero == 0)
            return 0;

        if (n == k)
        {
            if (zero == n)
                return 1;
            if (zero == 0)
                return 0;
            return -1;
        }

        int one = n - zero;
        int base = n - k;

        int ans = INT_MAX;

        // ---------- Try Odd Operations ----------
        if ((k % 2) == (zero % 2))
        {
            long long m = max(
                (zero + k - 1) / k,
                (one + base - 1) / base);

            if (m % 2 == 0)
                m++; // make odd

            ans = min(ans, (int)m);
        }

        // ---------- Try Even Operations ----------
        if (zero % 2 == 0)
        {
            long long m = max(
                (zero + k - 1) / k,
                (zero + base - 1) / base);

            if (m % 2 == 1)
                m++; // make even

            ans = min(ans, (int)m);
        }

        return ans == INT_MAX ? -1 : ans;
    }
};

File: Algorithms/src/3666. Minimum Operations to Equalize Binary String/Code/solution.go
func minOperations(s string, k int) int {
	n := len(s)
	zero := 0

	for i := 0; i < n; i++ {
		if s[i] == '0' {
			zero++
		}
	}

	if zero == 0 {
		return 0
	}

	if n == k {
		if zero == n {
			return 1
		}
		return -1
	}

	one := n - zero
	base := n - k

	ans := int(^uint(0) >> 1) // max int

	// Odd operations
	if (k%2) == (zero%2) {
		m := max(
			(zero+k-1)/k,
			(one+base-1)/base,
		)
		if m%2 == 0 {
			m++
		}
		if m < ans {
			ans = m
		}
	}

	// Even operations
	if zero%2 == 0 {
		m := max(
			(zero+k-1)/k,
			(zero+base-1)/base,
		)
		if m%2 == 1 {
			m++
		}
		if m < ans {
			ans = m
		}
	}

	if ans == int(^uint(0)>>1) {
		return -1
	}
	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}